README.md
# Algolia Search API Client for iOS and OS X


**&lt;Welcome Swift developers&gt;**

In June 2015, we release our [Swift API Client](https://github.com/algolia/algoliasearch-client-swift).

It is able to work with Swift and Objective-C. If you plan to use Swift in your project, please use it since we don't support Swift in our Objective-C API Client.

The Objective-C API Client is still supported and updated.

**&lt;/Welcome Swift developers&gt;**



[Algolia Search](http://www.algolia.com) is a hosted full-text, numerical, and faceted search engine capable of delivering realtime results from the first keystroke.

Our Objective-C client lets you easily use the [Algolia Search API](https://www.algolia.com/doc/rest_api) from your iOS & OS X applications. It wraps the [Algolia Search REST API](http://www.algolia.com/doc/rest_api).




[![Build Status](https://travis-ci.org/algolia/algoliasearch-client-objc.svg?branch=master)](https://travis-ci.org/algolia/algoliasearch-client-objc)
[![CocoaPods](https://img.shields.io/cocoapods/v/AlgoliaSearch-Client.svg)]()
[![CocoaPods](https://img.shields.io/cocoapods/p/AlgoliaSearch-Client.svg)]()
[![CocoaPods](https://img.shields.io/cocoapods/l/AlgoliaSearch-Client.svg)]()




Table of Contents
=================
**Getting Started**

1. [Setup](#setup)
1. [Quick Start](#quick-start)

1. [Online documentation](#documentation)
1. [Tutorials](#tutorials)


**Commands Reference**

1. [Add a new object](#add-a-new-object-to-the-index)
1. [Update an object](#update-an-existing-object-in-the-index)
1. [Search](#search)
1. [Search cache](#search-cache)
1. [Multiple queries](#multiple-queries)
1. [Get an object](#get-an-object)
1. [Delete an object](#delete-an-object)
1. [Delete by query](#delete-by-query)
1. [Index settings](#index-settings)
1. [List indices](#list-indices)
1. [Delete an index](#delete-an-index)
1. [Clear an index](#clear-an-index)
1. [Wait indexing](#wait-indexing)
1. [Batch writes](#batch-writes)
1. [Security / User API Keys](#security--user-api-keys)
1. [Copy or rename an index](#copy-or-rename-an-index)
1. [Backup / Retrieve all index content](#backup--retrieve-of-all-index-content)
1. [Logs](#logs)




Setup
-------------
To setup your project, follow these steps:




 1. [Download and add sources](https://github.com/algolia/algoliasearch-client-objc/archive/master.zip) to your project or use cocoapods by adding `pod 'AlgoliaSearch-Client', '~> 3.5'` in your Podfile (or `pod 'AlgoliaSearch-Client', '~> 2.0'` if your are using AFNetworking 1.x in your project)  or drop the source folder on your project (If you are not using a Podfile, you will also need to add [AFNetworking library](https://github.com/AFNetworking/AFNetworking) in your project).
 2. Add the `#import "ASAPIClient.h"` call to your project
 3. Initialize the client with your ApplicationID and API-Key. You can find all of them on [your Algolia account](http://www.algolia.com/users/edit).

```objc
ASAPIClient *apiClient =
    [ASAPIClient apiClientWithApplicationID:@"YourApplicationID" apiKey:@"YourAPIKey"];
```




Quick Start
-------------


In 30 seconds, this quick start tutorial will show you how to index and search objects.

Without any prior configuration, you can start indexing [500 contacts](https://github.com/algolia/algoliasearch-client-csharp/blob/master/contacts.json) in the ```contacts``` index using the following code:
```objc
// Load JSON file
NSString *jsonPath = [[NSBundle mainBundle] pathForResource:@"contacts" ofType:@"json"];
NSData* jsonData = [NSData dataWithContentsOfFile:jsonPath];
NSDictionary* dict = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:nil];
// Load all objects of json file in an index named "contacts"
ASRemoteIndex *index = [apiClient getIndex:@"contacts"];
[index addObjects:dict[@"objects"] success:nil failure:nil];
```

You can now search for contacts using firstname, lastname, company, etc. (even with typos):
```objc
// search by firstname
[index search:[ASQuery queryWithFullTextQuery:@"jimmie"] 
  success:^(ASRemoteIndex *index, ASQuery *query, NSDictionary *result) {
    NSLog(@"Result: %@", result);
} failure:nil];
// search a firstname with typo
[index search:[ASQuery queryWithFullTextQuery:@"jimie"] 
  success:^(ASRemoteIndex *index, ASQuery *query, NSDictionary *result) {
    NSLog(@"Result: %@", result);
} failure:nil];
// search for a company
[index search:[ASQuery queryWithFullTextQuery:@"california paint"] 
  success:^(ASRemoteIndex *index, ASQuery *query, NSDictionary *result) {
    NSLog(@"Result: %@", result);
} failure:nil];
// search for a firstname & company
[index search:[ASQuery queryWithFullTextQuery:@"jimmie paint"] 
  success:^(ASRemoteIndex *index, ASQuery *query, NSDictionary *result) {
    NSLog(@"Result: %@", result);
} failure:nil];
```

Settings can be customized to tune the search behavior. For example, you can add a custom sort by number of followers to the already great built-in relevance:
```objc
NSArray *customRanking = @[@"desc(followers)"];
NSDictionary *settings = @{@"customRanking": customRanking};
[index setSettings:settings success:nil
  failure:^(ASRemoteIndex *index, NSDictionary *settings, NSString *errorMessage) {
    NSLog(@"Error when applying settings: %@", errorMessage);
}];
```

You can also configure the list of attributes you want to index by order of importance (first = most important):
```objc
NSArray *customRanking = @[@"lastname", @"firstname", @"company", @"email", @"city", @"address"];
NSDictionary *settings = @{@"attributesToIndex": customRanking};
[index setSettings:settings success:nil
  failure:^(ASRemoteIndex *index, NSDictionary *settings, NSString *errorMessage) {
    NSLog(@"Error when applying settings: %@", errorMessage);
}];
```

Since the engine is designed to suggest results as you type, you'll generally search by prefix. In this case the order of attributes is very important to decide which hit is the best:
```objc
[index search:[ASQuery queryWithFullTextQuery:@"or"] 
  success:^(ASRemoteIndex *index, ASQuery *query, NSDictionary *result) {
    NSLog(@"Result: %@", result);
} failure:nil];

[index search:[ASQuery queryWithFullTextQuery:@"jim"] 
  success:^(ASRemoteIndex *index, ASQuery *query, NSDictionary *result) {
    NSLog(@"Result: %@", result);
} failure:nil];
```







Documentation
================
Check our [online documentation](http://www.algolia.com/doc/guides/objc):
 * [Initial Import](http://www.algolia.com/doc/guides/objc#InitialImport)
 * [Ranking &amp; Relevance](http://www.algolia.com/doc/guides/objc#RankingRelevance)
 * [Indexing](http://www.algolia.com/doc/guides/objc#Indexing)
 * [Search](http://www.algolia.com/doc/guides/objc#Search)
 * [Sorting](http://www.algolia.com/doc/guides/objc#Sorting)
 * [Filtering](http://www.algolia.com/doc/guides/objc#Filtering)
 * [Faceting](http://www.algolia.com/doc/guides/objc#Faceting)
 * [Geo-Search](http://www.algolia.com/doc/guides/objc#Geo-Search)
 * [Security](http://www.algolia.com/doc/guides/objc#Security)
 * [REST API](http://www.algolia.com/doc/rest)

Tutorials
================

Check out our [tutorials](http://www.algolia.com/doc/tutorials):
 * [Search bar with autocomplete menu](http://www.algolia.com/doc/tutorials/auto-complete)
 * [Search bar with multi category autocomplete menu](http://www.algolia.com/doc/tutorials/multi-auto-complete)
 * [Instant search result pages](http://www.algolia.com/doc/tutorials/instant-search)




Commands Reference
==================



Add a new object to the Index
-------------

Each entry in an index has a unique identifier called `objectID`. There are two ways to add en entry to the index:

 1. Using automatic `objectID` assignment. You will be able to access it in the answer.
 2. Supplying your own `objectID`.

You don't need to explicitly create an index, it will be automatically created the first time you add an object.
Objects are schema less so you don't need any configuration to start indexing. If you wish to configure things, the settings section provides details about advanced settings.

Example with automatic `objectID` assignment:

```objc
NSDictionary *newObject = @{@"firstname": @"Jimmie", @"lastname": @"Barninger"};
[index addObject:newObject 
  success:^(ASRemoteIndex *index, NSDictionary *object, NSDictionary *result) {
  	NSLog(@"Object ID: %@", result[@"objectID"]);
} failure:nil];
```

Example with manual `objectID` assignment:

```objc
NSDictionary *newObject = @{@"firstname": @"Jimmie", @"lastname": @"Barninger"};
[index addObject:newObject withObjectID:@"myID" 
  success:^(ASRemoteIndex *index, NSDictionary *object, NSString *objectID, NSDictionary *result) {
    NSLog(@"Object ID: %@", result[@"objectID"]);
} failure:nil];
```

Update an existing object in the Index
-------------

You have three options when updating an existing object:

 1. Replace all its attributes.
 2. Replace only some attributes.
 3. Apply an operation to some attributes.

Example on how to replace all attributes of an existing object:

```objc
NSDictionary *newObject = @{
	@"firstname": @"Jimmie",
    @"lastname": @"Barninger",
	@"city": @"New York"
};
[index saveObject:newObject objectID:@"myID" success:nil failure:nil];
```

You have many ways to update an object's attributes:

 1. Set the attribute value
 2. Add a string or number element to an array
 3. Remove an element from an array
 4. Add a string or number element to an array if it doesn't exist
 5. Increment an attribute
 6. Decrement an attribute

Example to update only the city attribute of an existing object:

```objc
NSDictionary *partialObject = @{@"city": @"San Francisco"};
[index partialUpdateObject:partialObject objectID:@"myID" success:nil failure:nil];
```

Example to add a tag:

```objc
NSDictionary *operation = @{
	@"value": @"MyTag",
	@"_operation": @"Add"
};
NSDictionary *partialObject = @{@"_tags": operation};
[index partialUpdateObject:partialObject objectID:@"myID" success:nil failure:nil];
```

Example to remove a tag:

```objc
NSDictionary *operation = @{
	@"value": @"MyTag",
	@"_operation": @"Remove"
};
NSDictionary *partialObject = @{@"_tags": operation};
[index partialUpdateObject:partialObject objectID:@"myID" success:nil failure:nil];
```

Example to add a tag if it doesn't exist:

```objc
NSDictionary *operation = @{
	@"value": @"MyTag",
	@"_operation": @"AddUnique"
};
NSDictionary *partialObject = @{@"_tags": operation};
[index partialUpdateObject:partialObject objectID:@"myID" success:nil failure:nil];
```

Example to increment a numeric value:

```objc
NSDictionary *operation = @{
	@"value": 42,
	@"_operation": @"Increment"
};
NSDictionary *partialObject = @{@"price": operation};
[index partialUpdateObject:partialObject objectID:@"myID" success:nil failure:nil];
```

Note: Here we are incrementing the value by `42`. To increment just by one, put
`value:1`.

Example to decrement a numeric value:

```objc
NSDictionary *operation = @{
	@"value": 42,
	@"_operation": @"Decrement"
};
NSDictionary *partialObject = @{@"price": operation};
[index partialUpdateObject:partialObject objectID:@"myID" success:nil failure:nil];
```

Note: Here we are decrementing the value by `42`. To decrement just by one, put
`value:1`.

Search
-------------


To perform a search, you only need to initialize the index and perform a call to the search function.

The search query allows only to retrieve 1000 hits, if you need to retrieve more than 1000 hits for seo, you can use [Backup / Retrieve all index content](#backup--retrieve-of-all-index-content)

You can use the following optional arguments on ASQuery class:

### Query Parameters

#### Full Text Search Parameters

 * **fullTextQuery**: (string) The instant search query string. All words of the query are interpreted as prefixes (for example "John Mc" will match "John Mccamey" and "Johnathan Mccamey"). If no query parameter is set all objects are retrieved.
 * **setQueryType**: Selects how the query words are interpreted. It can be one of the following values:
  * **prefixAll**: All query words are interpreted as prefixes.
  * **prefixLast**: Only the last word is interpreted as a prefix (default behavior).
  * **prefixNone**: No query word is interpreted as a prefix. This option is not recommended.
 * **setRemoveWordsIfNoResults**: This option is used to select a strategy in order to avoid having an empty result page. There are three different options:
  * **lastWords**: When a query does not return any results, the last word will be added as optional. The process is repeated with n-1 word, n-2 word, ... until there are results.
  * **firstWords**: When a query does not return any results, the first word will be added as optional. The process is repeated with second word, third word, ... until there are results.
  * **allOptional**: When a query does not return any results, a second trial will be made with all words as optional. This is equivalent to transforming the AND operand between query terms to an OR operand.
  * **none**: No specific processing is done when a query does not return any results (default behavior).
 * **minWordSizeForApprox1**: The minimum number of characters in a query word to accept one typo in this word.<br/>Defaults to 4.
 * **minWordSizeForApprox2**: The minimum number of characters in a query word to accept two typos in this word.<br/>Defaults to 8.
 * **setTypoTolerance**: This option allows you to control the number of typos in the result set:
  * **true**: The typo tolerance is enabled and all matching hits are retrieved (default behavior).
  * **false**: The typo tolerance is disabled. For example, if one result matches without typos, then all results with typos will be hidden.
  * **min**: Only keep results with the minimum number of typos.
  * **strict**: Hits matching with 2 typos are not retrieved if there are some matching without typos. This option is useful if you want to avoid false positives as much as possible.
 * **allowTyposOnNumericTokens**: If set to false, disables typo tolerance on numeric tokens (numbers). Defaults to true.
 * **ignorePlural**: If set to true, plural won't be considered as a typo. For example, car and cars will be considered as equals. Defaults to false.
 * **disableTypoToleranceOnAttributes** List of attributes on which you want to disable typo tolerance (must be a subset of the `attributesToIndex` index setting). Attributes are separated with a comma such as `"name,address"`. You can also use JSON string array encoding such as `encodeURIComponent("[\"name\",\"address\"]")`. By default, this list is empty.
 * **restrictSearchableAttributes** List of attributes you want to use for textual search (must be a subset of the `attributesToIndex` index setting). Attributes are separated with a comma such as `"name,address"`. You can also use JSON string array encoding such as `encodeURIComponent("[\"name\",\"address\"]")`. By default, all attributes specified in `attributesToIndex` settings are used to search.
 * **removeStopWords**: Remove stop words from query before executing it. Defaults to false. Contains stop words for 41 languages (Arabic, Armenian, Basque, Bengali, Brazilian, Bulgarian, Catalan, Chinese, Czech, Danish, Dutch, English, Finnish, French, Galician, German, Greek, Hindi, Hungarian, Indonesian, Irish, Italian, Japanese, Korean, Kurdish, Latvian, Lithuanian, Marathi, Norwegian, Persian, Polish, Portugese, Romanian, Russian, Slovak, Spanish, Swedish, Thai, Turkish, Ukranian, Urdu).
 * **advancedSyntax**: Enables the advanced query syntax. Defaults to 0 (false).
    * **Phrase query**: A phrase query defines a particular sequence of terms. A phrase query is built by Algolia's query parser for words surrounded by `"`. For example, `"search engine"` will retrieve records having `search` next to `engine` only. Typo tolerance is _disabled_ on phrase queries.
    * **Prohibit operator**: The prohibit operator excludes records that contain the term after the `-` symbol. For example, `search -engine` will retrieve records containing `search` but not `engine`.
 * **analytics**: If set to false, this query will not be taken into account in the analytics feature. Defaults to true.
 * **synonyms**: If set to false, this query will not use synonyms defined in the configuration. Defaults to true.
 * **replaceSynonymsInHighlight**: If set to false, words matched via synonym expansion will not be replaced by the matched synonym in the highlight results. Defaults to true.
 * **optionalWords**: An array of strings that contain the list of words that should be considered as optional when found in the query.

#### Pagination Parameters

 * **page**: (integer) Pagination parameter used to select the page to retrieve.<br/>Page is zero based and defaults to 0. Thus, to retrieve the 10th page you need to set `page=9`.
 * **hitsPerPage**: (integer) Pagination parameter used to select the number of hits per page. Defaults to 20.

#### Geo-search Parameters

 * **searchAroundLatitude:longitude:**: Search for entries around a given latitude/longitude.<br/>The maximum distance is automatically guessed depending of the density of the area but you also manually specify the maximum distance in meters with the **maxDist** parameter.<br/>At indexing, you should specify the geo location of an object with the `_geoloc` attribute in the form ` {"_geoloc":{"lat":48.853409, "lng":2.348800}} `.
 * **searchAroundLatitude:longitude:maxDist:precision**: Search for entries around a given latitude/longitude with a given precision for ranking. For example, if you set aroundPrecision=100, the distances will be considered by ranges of 100m, for example all distances 0 and 100m will be considered as identical for the "geo" ranking parameter.


 * **searchAroundLatitudeLongitudeViaIP**: Search for entries around the latitude/longitude automatically computed from user IP address.<br/>The radius is automatically guessed based on density but you can also specify it manually with the **maxDist** parameter (optional).<br/>At indexing, you should specify the geo location of an object with the `_geoloc` attribute in the form ` {"_geoloc":{"lat":48.853409, "lng":2.348800}} `.
 * **searchAroundLatitudeLlongitude:precision**: Search for entries around a latitude/longitude automatically computed from user IP address with a given precision for ranking. For example if you set precision=100, two objects that are a distance of less than 100 meters will be considered as identical for the "geo" ranking parameter.



ChangeLog
CHANGELOG

2016-02-05 3.6.4
- Added support of snippet ellipsis
	
2015-10-16 3.6.3
- Fixed ability to override index settings

2015-10-12 3.6.2
- Added remove stop words query parameter
- Added support of similar queries
- Added advanced syntax option
	
2015-09-30 3.6.1
 * Added support of multiple bounding box for geo-search
 * Added support of polygon for geo-search
 * Added support of automatic radius computation for geo-search
 * Added support of disableTypoToleranceOnAttributes        
	
2015-07-27 3.6.0
  * Add search cache (disabled by default)

2015-07-07 3.5.2
  * New browseFromCursor method
  * Add analyticsTags to Query

2015-07-01 3.5.1
  * Fix grouping

2015-07-01 3.5.0
  * ASQuery implements protocol NSCopying
  * Fix various bugs

2015-06-30 3.4.6
  * Added support of grouping (distinct=3 for example keep the 3 best hits for a distinct key)

2015-06-16 3.4.5
  * New browse method that takes a ASQuery object
  * Remove DSN flag

2015-05-04 3.4.4
  * add new parameter on the Query: setMinProximity & setHighlightingTags

2015-05-04 3.4.3
  * Add new methods to add/update api key
  * Add batch method to target multiple indices
  * Add strategy parameter for the multipleQueries
  * Add new method to generate secured api key from query parameters

2015-04-22 3.4.2
  * Fix issue with geo search

2015-04-10 3.4.1
  * Return the error message from the backend instead of a static message for 4XX

2015-04-10 3.4.0
  * Add new retry logic

2014-11-30 3.2.0
  * Switch .io to .net

2014-11-07 3.1.23
  * Add DSN flag

2014-10-27  3.1.22
  * Add extra optionalWordsMinimumMatched query parameter


src/ASQuery.h
//
//  Copyright (c) 2013 Algolia
//  http://www.algolia.com/
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

#import <Foundation/Foundation.h>

/**
 * Describes all parameters of search query.
 */
@interface ASQuery : NSObject <NSCopying>

/**
 * Initialize query with a full text query string
 */
+(instancetype) queryWithFullTextQuery:(NSString*)fullTextQuery;

/**
 * Initialize an empty query
 */
-(instancetype) init;

/**
 * Initialize query with a full text query string
 */
-(instancetype) initWithFullTextQuery:(NSString*)fullTextQuery;

-(instancetype) copyWithZone:(NSZone*)zone;

/**
 *  Search for entries around a given latitude/longitude with an automatic radius
 */
-(ASQuery*) searchAroundLatitude:(float)latitude longitude:(float)longitude;

/**
 *  Search for entries around a given latitude/longitude.
 *
 *  @param maxDist set the maximum distance in meters.
 *  Note: at indexing, geoloc of an object should be set with _geoloc attribute containing lat and lng attributes (for example {"_geoloc":{"lat":48.853409, "lng":2.348800}})
 */
-(ASQuery*) searchAroundLatitude:(float)latitude longitude:(float)longitude maxDist:(NSUInteger)maxDist;

/**
 *  Search for entries around a given latitude/longitude.
 *
 *  @param maxDist set the maximum distance in meters.
 *  @param precision set the precision for ranking (for example if you set precision=100, two objects that are distant of less than 100m will be considered as identical for "geo" ranking parameter).
 *  Note: at indexing, geoloc of an object should be set with _geoloc attribute containing lat and lng attributes (for example {"_geoloc":{"lat":48.853409, "lng":2.348800}})
 */
-(ASQuery*) searchAroundLatitude:(float)latitude longitude:(float)longitude maxDist:(NSUInteger)maxDist precision:(NSUInteger)precision;

/**
 *  Search for entries around a given latitude/longitude (using IP geolocation) with an automatic radius
 */
-(ASQuery*) searchAroundLatitudeLongitudeViaIP;

/**
 *  Search for entries around a given latitude/longitude (using IP geolocation)
 *
 *  @param maxDist set the maximum distance in meters.
 *  Note: at indexing, geoloc of an object should be set with _geoloc attribute containing lat and lng attributes (for example {"_geoloc":{"lat":48.853409, "lng":2.348800}})
 */
-(ASQuery*) searchAroundLatitudeLongitudeViaIP:(NSUInteger)maxDist;

/**
 *  Search for entries around a given latitude/longitude (using IP geolocation)
 *
 *  @param maxDist set the maximum distance in meters.
 *  @param precision set the precision for ranking (for example if you set precision=100, two objects that are distant of less than 100m will be considered as identical for "geo" ranking parameter).
 *  Note: at indexing, geoloc of an object should be set with _geoloc attribute containing lat and lng attributes (for example {"_geoloc":{"lat":48.853409, "lng":2.348800}})
 */
-(ASQuery*) searchAroundLatitudeLongitudeViaIP:(NSUInteger)maxDist precision:(NSUInteger)precision;


/**
 * Search for entries inside a given area defined by the two extreme points of a rectangle.
 * At indexing, you should specify geoloc of an object with the _geoloc attribute (in the form "_geoloc":{"lat":48.853409, "lng":2.348800} or 
 * "_geoloc":[{"lat":48.853409, "lng":2.348800},{"lat":48.547456, "lng":2.972075}] if you have several geo-locations in your record).
 * 
 * You can use several bounding boxes (OR) by calling this method several times.
 */
-(ASQuery*) searchInsideBoundingBoxWithLatitudeP1:(float)latitudeP1 longitudeP1:(float)longitudeP1 latitudeP2:(float)latitudeP2 longitudeP2:(float)longitudeP2;

/**
 * Add a point to the polygon of geo-search (requires a minimum of three points to define a valid polygon)
 * At indexing, you should specify geoloc of an object with the _geoloc attribute (in the form "_geoloc":{"lat":48.853409, "lng":2.348800} or 
 * "_geoloc":[{"lat":48.853409, "lng":2.348800},{"lat":48.547456, "lng":2.972075}] if you have several geo-locations in your record).
 */
-(ASQuery*) addInsidePolygon:(float)latitude longitude:(float)longitude;

/**
 * Return the final query string used in URL.
 */
-(NSString*) buildURL;

/**
 * Select how the query words are interpreted:
 * "prefixAll": all query words are interpreted as prefixes,
 * "prefixLast": only the last word is interpreted as a prefix (default behavior),
 * "prefixNone": no query word is interpreted as a prefix. This option is not recommended.
 */
@property (nonatomic) NSString            *queryType;

/**
 * Set full text search of similar query (like this)
 */
@property (nonatomic) NSString            *similarQuery;

/**
 * Select the strategy to avoid having an empty result page.
 * "None": No specific processing is done when a query does not return any result,
 * "LastWords": when a query does not return any result, the final word will be removed until there is results,
 * "FirstWords": when a query does not return any result, the first word will be removed until there is results.
 * "allOptional": When a query does not return any result, a second trial will be made with all words as optional (which is equivalent to transforming the AND operand between query terms in a OR operand)
 */
@property (nonatomic) NSString            *removeWordsIfNoResult;

/**
 * Specify the list of attribute names to retrieve.
 * By default all attributes are retrieved.
 */
@property (nonatomic) NSArray             *attributesToRetrieve;

/**
 * Specify the list of attribute names to highlight.
 * By default indexed attributes are highlighted.
 */
@property (nonatomic) NSArray             *attributesToHighlight;

/**
 * Specify the List of attributes on which you want to disable typo tolerance (must be a subset of the attributesToIndex index setting).
 * By default this list is empty
 */
@property (nonatomic) NSArray             *disableTypoToleranceOnAttributes;

/**
 * Specify the list of attributes to snippet alongside the number of words to return 
 * (syntax is 'attributeName:nbWords'). 
 * Attributes are separated by a comma (Example: "attributesToSnippet=name:10,content:10").
 * By default no snippet is computed.
*/
@property (nonatomic) NSArray             *attributesToSnippet;

/**
 * Filter the query by a set of tags. You can AND tags by separating them by commas. To OR tags, you must add parentheses. For example tag1,(tag2,tag3) means tag1 AND (tag2 OR tag3).
 * At indexing, tags should be added in the _tags attribute of objects (for example {"_tags":["tag1","tag2"]} )
 */
@property (nonatomic) NSString            *tagFilters;

/**
 * Add a list of numeric filters separated by a comma.
 * The syntax of one filter is `attributeName` followed by `operand` followed by `value. Supported operands are `<`, `<=`, `=`, `>` and `>=`.
 * You can have multiple conditions on one attribute like for example `numerics=price>100,price<1000`.
 */
@property (nonatomic) NSString            *numericFilters;

/**
 * Set the full text query.
 */
@property (nonatomic) NSString            *fullTextQuery;

/**
 * Specify the minimum number of characters in a query word to accept one typo in this word.
 * Defaults to 3.
 */
@property (nonatomic) NSUInteger           minWordSizeForApprox1;

/**
 * Specify the minimum number of characters in a query word to accept two typos in this word.
 * Defaults to 7.
 */
@property (nonatomic) NSUInteger           minWordSizeForApprox2;

/**
 * Set the page to retrieve (zero base). Defaults to 0.
 */
@property (nonatomic) NSUInteger           page;

/**
 *  Set the number of hits per page. Defaults to 10.
 */
@property (nonatomic) NSUInteger           hitsPerPage;

/**
 * Configure the precision of the proximity ranking criterion. By default, the minimum (and best) proximity value distance between 2 matching words is 1. Setting it to 2 (or 3) would allow 1 (or 2) words to be found between the matching words without degrading the proximity ranking value.
 * Considering the query "javascript framework", if you set minProximity=2 the records "JavaScript framework" and "JavaScript charting framework" will get the same proximity score, even if the second one contains a word between the 2 matching words. Default to 1.
 */
@property (nonatomic) NSUInteger           minProximity;

/**
 * if set, the result hits will contain ranking information in _rankingInfo attribute.
 */
@property BOOL                             getRankingInfo;

/**
 * If set to YES, plural won't be considered as a typo (for example car/cars will be considered as equals). Default to NO.
 */
@property BOOL                             ignorePlural;

/**
 * This option allow to control the number of typo in the results set.
 */
@property (nonatomic) NSString            *typoTolerance;

/**
 *  If set to false, disable typo-tolerance on numeric tokens. Default to true.
 */
@property BOOL                             typosOnNumericTokens;

/**
 * If set to false, disable this query won't appear in the analytics. Default to true.
 */
@property BOOL                             analytics;

/**
 * If set to false, this query will not use synonyms defined in configuration. Default to true.
 */
@property BOOL                             synonyms;

/**
 * If set to false, words matched via synonyms expansion will not be replaced by the matched synonym in highlight result. Default to true.
 */
@property BOOL                             replaceSynonyms;

/**
 *
 * Enable the distinct feature (disabled by default) if the attributeForDistinct index setting is set.
 * This feature is similar to the SQL "distinct" keyword: when enabled in a query with the distinct=1 parameter,
 * all hits containing a duplicate value for the attributeForDistinct attribute are removed from results.
 * For example, if the chosen attribute is show_name and several hits have the same value for show_name,
 * then only the best one is kept and others are removed.
 * Specify the maximum number of hits to keep for each distinct value.
 */
@property (nonatomic) NSUInteger           distinct;

/**
  * Set the list of words that should be considered as optional when found in the query (array of NSString).
  */
@property (nonatomic) NSArray             *optionalWords;

/**
 *  Set the minimum number of optional words that need to match
 */
@property (nonatomic) NSUInteger           optionalWordsMinimumMatched;

/**
 * Filter the query with numeric, facet or/and tag filters. The syntax is a SQL like syntax, you can use the OR and AND keywords.
 * The syntax for the underlying numeric, facet and tag filters is the same than in the other filters:
 * available=1 AND (category:Book OR NOT category:Ebook) AND public
 * date: 1441745506 TO 1441755506 AND inStock > 0 AND author:"John Doe"
 * The list of keywords is:
 * OR: create a disjunctive filter between two filters.
 * AND: create a conjunctive filter between two filters.
 * TO: used to specify a range for a numeric filter.
 * NOT: used to negate a filter. The syntax with the ‘-‘ isn’t allowed.
 */
@property (nonatomic) NSString             *filters;

/**
 * Filter the query by a list of facets. Each facet is encoded as `attributeName:value`. For example: ["category:Book","author:John%20Doe"].
 */
@property (nonatomic) NSArray             *facetFilters;

/**
 * Filter the query by a list of facets encoded as one string by example "(category:Book,author:John)"
 */
@property (nonatomic) NSString            *facetFiltersRaw;

/**
 * List of object attributes that you want to use for faceting.
 * Only attributes that have been added in **attributesForFaceting** index setting can be used in this parameter. 
 * You can also use `*` to perform faceting on all attributes specified in **attributesForFaceting**.
 */
@property (nonatomic) NSArray             *facets;

/**
 * List of object attributes you want to use for textual search (must be a subset of the attributesToIndex 
 * index setting). Attributes are separated with a comma (for example @"name,address").
 * You can also use a JSON string array encoding (for example encodeURIComponent("[\"name\",\"address\"]")).
 * By default, all attributes specified in attributesToIndex settings are used to search.
 */
@property (nonatomic) NSString            *restrictSearchableAttributes;

/**
 * Specify the string that is inserted before the highlighted parts in the query result (default to "<em>").
 */
@property (nonatomic) NSString            *highlightPreTag;

/**
 * Specify the string that is inserted after the highlighted parts in the query result (default to "</em>").
 */
@property (nonatomic) NSString            *highlightPostTag;

/**
 * Specify the string that is used as an ellipsis indicator when a snippet
 * is truncated (defaults to the empty string).
 */
@property (nonatomic) NSString            *snippetEllipsisText;

/**
 * Contains insideBoundingBox query (you should use searchInsideBoundingBox selector to set it)
 */
@property (nonatomic) NSString            *insideBoundingBox;

/**
 * Contains insidePolygon query (you should use addInsidePolygon selector to set it)
 */
@property (nonatomic) NSString            *insidePolygon;

/**
 * Contains aroundLatLong query (you should use searchAroundLatitude:longitude:maxDist selector to set it)
 */
@property (nonatomic) NSString            *aroundLatLong;

/**
 * Change the radius or around latitude/longitude query
 */
@property (nonatomic) NSUInteger           aroundRadius;

/**
 * Change the precision or around latitude/longitude query
 */
@property (nonatomic) NSUInteger           aroundPrecision;

/**
 * Tags can be used in the Analytics to analyze a subset of searches only. Comma-separated string list like @[@"ios", @"web"]
 */
@property (nonatomic) NSArray            *analyticsTags;

/**
 * If set to YES use geolocation via client IP instead of passing a latitude/longitude manually
 */
@property BOOL                             aroundLatLongViaIP;

/**
 * If set to YES,  the advanced query syntax will be availabel. Default to false.
 */
@property BOOL                             advancedSyntax;

/**
 * If set to YES, enable removal of stop words
 */
@property BOOL                             removeStopWords;

/**
 *
 */
@property (nonatomic) NSString             *userToken;

/**
 *
 */
@property (nonatomic) NSString             *referers;

/**
 * Define the maximum number of values returns foreach facet
 */
@property (nonatomic) NSUInteger           maxValuesPerFacet;

@end
